src/mame/drivers/magicard.cpp

```diff
@@ -350,313 +551,616 @@ TODO: check this register,doesn't seem to be 100% correct.
+	map(0x00680000, 0x006ffbff).rom().region("maincpu", 0); // boot vectors point here
+	map(0x00800000, 0x0087fbff).rom().region("maincpu", 0); // boot vectors point here
+	map(0x001fff80, 0x001fffbf).ram(); //DRAM I/O, not accessed by this game, CD buffer?
+	map(0x00400000, 0x0040ffff).ram().share("nvram");
+	map(0x00411000, 0x00411001).portr("SW0");
+	map(0x00412346, 0x00412347).portr("SW1");
+	map(0x00413000, 0x00413001).w( FUNC(hotslots_state::output_w));
 	map(0x00414001, 0x00414001).w("ramdac", FUNC(ramdac_device::index_w));
 	map(0x00414003, 0x00414003).w("ramdac", FUNC(ramdac_device::pal_w));
 	map(0x00414005, 0x00414005).w("ramdac", FUNC(ramdac_device::mask_w));
 	map(0x00414007, 0x00414007).w("ramdac", FUNC(ramdac_device::index_w));
 	map(0x00415003, 0x00415003).r("ramdac", FUNC(ramdac_device::pal_r));
 	map(0x00416001, 0x00416001).w("ssg", FUNC(ymz284_device::data_w));
 	map(0x00417001, 0x00417001).w("ssg", FUNC(ymz284_device::address_w));
+	map(0x00418000, 0x00418020).rw("rtc", FUNC(rtc72421_device::read), FUNC(rtc72421_device::write)).umask16(0x00ff);
 }
 
+void hotslots_state::hotslots_map(address_map &map)
+{
+	hotslots_map_base(map);
+	map(0x00419566, 0x00419567).rw( FUNC(hotslots_state::read_ds1207_ds2401), FUNC(hotslots_state::write_ds1207_ds2401)).umask16(0x00ff);
+}
 
-/*********************************************
-*                Input Ports                 *
-*********************************************/
+void hotslots_state::puzzleme_map(address_map &map)
+{
+	hotslots_map_base(map);
+}
+
+/*************************
+*      Input ports       *
+*************************/
 
 static INPUT_PORTS_START( magicard )
-	PORT_START("SYSTEM")
-	PORT_DIPNAME( 0x01, 0x01, "SYSTEM0" )
-	PORT_DIPSETTING(    0x01, DEF_STR( Off ) )
-	PORT_DIPSETTING(    0x00, DEF_STR( On ) )
-	PORT_DIPNAME( 0x02, 0x02, DEF_STR( Unknown ) )
-	PORT_DIPSETTING(    0x02, DEF_STR( Off ) )
+	PORT_START("SW0")
+
+	PORT_BIT( 0x01, IP_ACTIVE_LOW, IPT_COIN1 )
+
+	PORT_BIT( 0x02, IP_ACTIVE_LOW, IPT_COIN3 ) PORT_NAME("Remote 2")
+	PORT_BIT( 0x04, IP_ACTIVE_LOW, IPT_COIN4 ) PORT_NAME("Remote 1")
+
+	PORT_BIT( 0x08, IP_ACTIVE_LOW, IPT_COIN2 )
+
+	PORT_BIT( 0x10, IP_ACTIVE_LOW, IPT_POKER_HOLD1 ) PORT_NAME("Hold 1")
+
+	PORT_BIT( 0x20, IP_ACTIVE_LOW, IPT_BUTTON1 ) PORT_NAME("Bet/Clear/Collect")
+
+	PORT_BIT( 0x40, IP_ACTIVE_LOW, IPT_POKER_HOLD5 ) PORT_NAME("Hold 5")
+
+	PORT_BIT( 0x80, IP_ACTIVE_LOW, IPT_START1 )
+
+	PORT_BIT( 0x100, IP_ACTIVE_LOW, IPT_SERVICE2 ) PORT_NAME("Rental Book Keeping") PORT_TOGGLE
+	PORT_BIT( 0x200, IP_ACTIVE_LOW, IPT_SERVICE1 ) PORT_NAME("Owner Book Keeping")
+
+	PORT_BIT( 0x0400, IP_ACTIVE_LOW, IPT_POKER_HOLD4 ) PORT_NAME("Hold 4")
+	PORT_BIT( 0x0800, IP_ACTIVE_LOW, IPT_POKER_HOLD2 ) PORT_NAME("Hold 2")
+	PORT_BIT( 0x1000, IP_ACTIVE_LOW, IPT_POKER_HOLD3 ) PORT_NAME("Hold 3")
+
+	PORT_BIT( 0x2000, IP_ACTIVE_LOW, IPT_BUTTON4 ) PORT_NAME("Pay Out") PORT_CODE(KEYCODE_Q)
```

Comment:
This should probably either use `IPT_GAMBLE_KEYOUT` if it’s an attendant control, or `IPT_GAMBLE_PAYOUT` if it’s a player control.